x-pack/plugins/apm/public/hooks/use_table_pagination.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CriteriaWithPagination,
+  EuiTableSortingType,
+  Pagination,
+} from '@elastic/eui';
+import { orderBy } from 'lodash';
+import { useMemo, useState } from 'react';
+import { useHistory } from 'react-router-dom';
+import uuid from 'uuid';
+import { fromQuery, toQuery } from '../components/shared/links/url_helpers';
+import { useApmParams } from './use_apm_params';
+
+interface Initials<T extends any[]> {
+  // totalItemCount is not necessary here
+  initialPagination: Omit<Pagination, 'totalItemCount'>;
+  initialSort?: EuiTableSortingType<T[0]>;
+}
+
+type Props<T extends any[]> = {
+  items: T;
+  urlState?: boolean;
+} & Initials<T>;
+
+const PAGE_SIZE_OPTIONS = [10, 25, 50];
+
+export function useTablePagination<T extends any[]>({
+  items,
+  initialPagination,
+  initialSort = {},
+  urlState = false,
+}: Props<T>): {
+  onTableChange: (criteriaWithPagination: CriteriaWithPagination<T[0]>) => void;
+  tableSort?: EuiTableSortingType<T[0]>;
+  tablePagination: Pagination;
+  tableItems: any[];
+  totalItems: number;
+  requestId: string;
+} {
+  const { query } = useApmParams('/*');
+
+  const { tableOptionsUrl, onTableChangeUrl } = useTablePaginationUrl({
+    initialPagination,
+    initialSort,
+  });
+
+  const [tableOptionsState, setTableOptionsState] = useState<
+    CriteriaWithPagination<T[0]>
+  >({
+    page: {
+      index: initialPagination.pageIndex,
+      size: initialPagination.pageSize,
+    },
+    sort: initialSort.sort,
+  });
+
+  const tableOptions = useMemo(
+    () => (urlState ? tableOptionsUrl : tableOptionsState),
+    [urlState, tableOptionsUrl, tableOptionsState]
+  );
+
+  const offset = 'offset' in query ? query.offset : undefined;
+  const comparisonEnabled =
+    'comparisonEnabled' in query ? query.comparisonEnabled : undefined;
+
+  const { tableItems, requestId } = useMemo(
+    () => {
+      return {
+        tableItems: orderBy(
+          items,
+          tableOptions.sort?.field,
+          tableOptions.sort?.direction
+        ).slice(
+          tableOptions.page.index * tableOptions.page.size,
+          (tableOptions.page.index + 1) * tableOptions.page.size
+        ),
+        // Generate a new id everytime the table options are changed
+        requestId: uuid(),
```

Comment:
That's the thing, the hook controls the pagination, so when we paginate a new requestId has to be generated in order to trigger the comparison API call. 

If the caller is responsible to handle it, the `useFetcher` dependency will have to list the pagination and sort instructions.

Reply:
That doesn't work for the first time it loads the page. That is because the comparison API call only receives the `requestId` as a parameter, as it should. If I move the requestId to a state, that won't trigger the comparison API call after the main API returned, because requestId will be the same. Is that clear?